pkg/utils/test/testf/testf_assertions.go

```
@@ -0,0 +1,167 @@
+package testf
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+	"time"
+
+	"github.com/onsi/gomega"
+	"github.com/onsi/gomega/types"
+)
+
+type Mode int
+
+const (
+	eventually Mode = iota
+	consistently
+)
+
+type EventuallyValue[T any] struct {
+	ctx context.Context
+	g   *gomega.WithT
+	f   func(context.Context) (T, error)
+}
+
+func (e *EventuallyValue[T]) Get() (T, error) {
```

Can you describe intention of this code? Because from what I understood so far, the check should be errors.Is (no difference in code flow, but different intentions) and if Unwrap returns nil, it is basically panic()

---

The `EventuallyValue` struct is meant to wrap gomega calls so that you can:
- perform async assertions using eventually or consistently and then get the result if the assertion succeed
- directly get the value without assertions (this is something that will be added later)

So you can do something like:

```go
dsci, err := g.Get(gvk.DSCInitialization, c.DSCIName).Get()
g.Expect(err).ShouldNot(HaveOccurred())

g.Get(gvk.Secret, types.NamespacedName{...}).Eventually().Should(And(
  jq.Match(`.type == "%s"`, is.Type),
  jq.Match(`(.data."tls.crt" | @base64d) == "%s"`, is.Data["tls.crt"]),
  jq.Match(`(.data."tls.key" | @base64d) == "%s"`, is.Data["tls.key"]),
))
```

Given the function retrieval can be invoked by directly or by using gomega, it may return `PollingSignalError` in case a non recoverable error is found and if so, return the wrapped error if any. I cannot use `error.Is` because that method would check for the error equality, whereas I'm only interested about the type.

Not sure if I missed anything.

---

Yes, I'm wrong about As vs Is.

But the second part. Any case when nil error can be wrapped and it's a normal situation and original error then must be returned? I see it more as
```
if errors.As(err, &pse) {
	err = errors.Unwrap(err)
}
return v, err
```

---

I'm not sure about this because the `PollingSignalError` is not guaranteed to always wrap an error so it would be perfectly fine to have something like:

```go
if len(foo) == 0{
    return gomega.StopTrying("should not be empty")
}
```

in this case we would swallow the error if we always get the wrapped one (in such case `nil`)